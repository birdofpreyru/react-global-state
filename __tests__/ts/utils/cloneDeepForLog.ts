import type * as LodashM from 'lodash';
import mockdate from 'mockdate';

const mockCloneDeepDelay = 500;

const consoleWarnings: unknown[][] = [];

console.warn = (...data: unknown[]) => {
  consoleWarnings.push(data);
};

jest.mock<typeof LodashM>('lodash-es', () => {
  const real = jest.requireActual<typeof LodashM>('lodash-es');
  return {
    ...real,
    cloneDeep<T>(value: T) {
      mockdate.set(Date.now() + mockCloneDeepDelay);
      return real.cloneDeep(value);
    },
  };
});

test('key timeout works as expected', async () => {
  // eslint-disable-next-line import/dynamic-import-chunkname
  const { cloneDeepForLog } = await import('src/utils');

  const oA = { data: { key: 'value' } };
  let oB = cloneDeepForLog(oA, '2');
  oA.data.key = 'value-2';
  expect(oB.data.key).toBe('value');
  expect(consoleWarnings).toMatchSnapshot();

  oB = cloneDeepForLog(oA, '2');
  oA.data.key = 'value-3';
  expect(oB.data.key).toBe('value-3');
  expect(consoleWarnings).toMatchSnapshot();

  oB = cloneDeepForLog(oA, '3');
  oA.data.key = 'value-4';
  expect(oB.data.key).toBe('value-3');
  expect(consoleWarnings).toMatchSnapshot();
});
