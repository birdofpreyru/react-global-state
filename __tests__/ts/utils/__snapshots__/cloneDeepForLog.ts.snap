// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`key timeout works as expected 1`] = `
[
  [
    "500ms spent to clone the logged value (key "2").",
  ],
]
`;

exports[`key timeout works as expected 2`] = `
[
  [
    "500ms spent to clone the logged value (key "2").",
  ],
  [
    "The logged value won't be clonned (key "2").",
  ],
]
`;

exports[`key timeout works as expected 3`] = `
[
  [
    "500ms spent to clone the logged value (key "2").",
  ],
  [
    "The logged value won't be clonned (key "2").",
  ],
  [
    "500ms spent to clone the logged value (key "3").",
  ],
]
`;
